padertorch/ops/stft_module.py

```diff
@@ -0,0 +1,270 @@
+                                      'for later implementations')
+        self.complex_representation = complex_representation
+        self.size = size
+        self.shift = shift
+        self.window_length = window_length if window_length is not None \
+            else size
+
+        window = _get_window(
+            window=window,
+            symmetric_window=symmetric_window,
+            window_length=self.window_length,
+        )
+        assert fading in [None, True, False, 'full', 'half'], fading
+        self.fading = fading
+        self.pad = pad
+        self.stft_kernel = get_stft_kernel(size, window)
+        self.istft_kernel_real, self.istft_kernel_imag = get_istft_kernel(
+            size, shift, window)
+
+    def __call__(self, inputs):
+        """
+        Args:
+            inputs: shape: [..., T], T is #samples
+            num_samples, list or tensor of #samples
+
+        >>> mixture = torch.rand((2, 6, 203))
+        >>> torch_stft = STFT(512, 20, window_length=40,\
+                              complex_representation='concat')
+        >>> torch_stft_out = torch_stft(mixture)
+        >>> torch_stft_out.shape
+        torch.Size([2, 6, 12, 514])
+        >>> from paderbox.transform import stft
+        >>> stft_out = stft(mixture.numpy(), 512, 20, window_length=40)
+        >>> stft_signal = np.concatenate(\
+                [np.real(stft_out), np.imag(stft_out)], axis=-1)
+        >>> np.testing.assert_allclose(torch_stft_out, stft_signal, atol=1e-5)
+
+        """
+
+        org_shape = inputs.shape
+        stride = self.shift
+        length = self.window_length
+        x = inputs.view((-1, org_shape[-1]))
+        # Pad with zeros to have enough samples for the window function to fade.
+        assert self.fading in [None, True, False, 'full', 'half'], self.fading
+        if self.fading not in [False, None]:
+            if self.fading == 'half':
+                pad_width= (
+                    (self.window_length - self.shift) // 2,
+                    ceil((self.window_length - self.shift) / 2)
+                )
+            else:
+                pad_width = self.window_length - self.shift
+                pad_width = (pad_width, pad_width)
+            x = F.pad(x, pad_width, mode='constant')
+
+        if self.pad:
+            if x.shape[-1] < length:
+                pad_size = length - x.shape[-1]
+                x = F.pad(x, (0, pad_size))
+            elif stride != 1 and (x.shape[-1] + stride - length) % stride != 0:
+                pad_size = stride - ((x.shape[-1] + stride - length) % stride)
+                x = F.pad(x, (0, pad_size))
+
+        x = torch.unsqueeze(x, 1) # [..., 1, T]
+        weights = self.stft_kernel.to(x)
+        encoded = F.conv1d(x, weight=weights, stride=stride)
+
+        encoded = encoded.view(*org_shape[:-1], *encoded.shape[-2:])
+        encoded = rearrange(encoded, '... feat frames -> ... frames feat')
```

Comment:
Can you do the `view` `rearrange` in a single `rearrange`?
Something like `rearrange(encoded, '... (feat frames) -> ... frames feat', feat=..., frames=...)` should work.